File: Design1-Q2.py
class SplStack:
    def __int__(self):
        self.stack=[]
        self.min=[]
        self.size=0
        self._top=-1
        self.maxsize=10


     def isEmpty(self):
         return self._top=-1

     def isFull(self):
         if self.size== self.maxsize :
             return True
         else:
             return False

    def push(self, x):
        if self.size==0:
          self.min.append(x)
        else:
            if x <= self.min[-1]:
                self.min.append(x)
        self.stack.append(x)
        self.size+=1

    def pop(self):
        temp=self.stack.pop()
        self.size-=1
        if temp==self.min[-1] :
          self.min.pop()

    def getMin(self):
        return self.min[-1]

File: design-1q2.java
//design-1q2 - minstack
//LC 155
// we will keep 2 stacks and every time a new element is inserted in the main stack we will compare with the min in min stack


class MinStack {
    private Stack<Integer> st;
    private Stack<Integer> minSt;
    int min;
    public MinStack() {
        min = Integer.MAX_VALUE;
        st = new Stack<>();
        minSt = new Stack<>();
        minSt.push(min);
        
    }
    // Overall Time complexity - O(1)
    public void push(int val) {
        min = Math.min(min, val);
        st.push(val);
        minSt.push(min);
    }
    // Overall Time complexity - O(1)
    public void pop() {
        st.pop();
        minSt.pop();
        min = minSt.peek();
    }
    // Overall Time complexity - O(1)
    public int top() {
        return st.peek();
    }
    // Overall Time complexity - O(1)
    public int getMin() {
        return minSt.peek();
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

